### PR TITLE
Fix #3939: Add Screen Time support.

### LINKED ISSUE #3939: Implement screentime
### Description: 
See this API https://developer.apple.com/documentation/screentime/stwebpagecontroller
It may be possible to block restricted pages based on screentime limits

### Steps to Reproduce 
  1.
  2.
  3.

**Actual result:** 

**Expected result:**

**Reproduces how often:** [Easily reproduced, Intermittent Issue]

**Brave Version:** 

**Device details:** 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### PR DESCRIPTION
**Pull request by iccub**
## Summary of Changes

Old PR by our contributor here https://github.com/brave/brave-ios/pull/7639

This pull request fixes #3939 
Security review https://github.com/brave/reviews/issues/1383

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:

1. Verify that when you stay on a website for a bit it is tracked in Screen Time section in iOS settings.
2. Verify that after setting a time limit for a website, after you reach the limit you are not able to access it anymore.
3. Verify that websites are not tracked in private browsing mode

Disabled tests:
1. Turn off the feature - verify that websites are not tracked again
2. Turn off the feature - verify that the website where you reached the time limit can be accessed

## Screenshots:

All the in-website UI is Apple controlled.
Zrzut ekranu 2023-08-29 o 13 40 11

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ScreenTime.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  func updateScreenTimeUrl(_ url: URL?) {
+    guard let url = url, (url.scheme == "http" || url.scheme == "https") else {
+      screenTimeViewController?.url = URL(string: "https://about:blank")
+      return
+    }
+    
+    screenTimeViewController?.url = url
+  }
+  
+  func recordScreentimeUsage(for tab: Tab) {
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ScreenTime.swift]: `Screentime` → `ScreenTime` for consistency

File: Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+TabManagerDelegate.swift [hunk 2]
```diff
@@ -41,6 +41,7 @@ extension BrowserViewController: TabManagerDelegate {
         })
       
       updateURLBar()
+      recordScreentimeUsage(for: tab)
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+TabManagerDelegate.swift]: `Screentime` → `ScreenTime` for consistency

File: Sources/Brave/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -3152,6 +3172,19 @@ extension BrowserViewController: PreferencesObserver {
       syncPlaylistFolders()
     case Preferences.NewTabPage.backgroundSponsoredImages.key:
       recordAdsUsageType()
+    case Preferences.Privacy.screenTimeEnabled.key:
+      if Preferences.Privacy.screenTimeEnabled.value {
+        screenTimeViewController = STWebpageController()
+        if let tab = tabManager.selectedTab {
+          recordScreentimeUsage(for: tab)
+        }
+        updateViewConstraints()
```
Comment (Brandon-T) [inline Sources/Brave/Frontend/Browser/BrowserViewController.swift]: Is this actually okay to call manually or do you have to call `setNeedsLayout` and `layoutIfNeeded` instead?

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: Package.swift [hunk 4]
```diff
@@ -51,7 +51,7 @@ var package = Package(
     .package(url: "https://github.com/SDWebImage/SDWebImage", exact: "5.10.3"),
     .package(url: "https://github.com/SDWebImage/SDWebImageSwiftUI", from: "2.2.0"),
     .package(url: "https://github.com/nmdias/FeedKit", from: "9.1.2"),
-    .package(url: "https://github.com/brave/PanModal", revision: "e4c07f8e6c5df937051fabc47e1e92901e1d068b"),
+    .package(url: "https://github.com/brave/PanModal", revision: "e67e9eff53c05f19b41bbb2ca7d27ff5859a586c"),
```
Comment (soner-yuksel) [inline Package.swift:54]: Why did we need to update PanModal ?

Comment (iccub) [reply Package.swift:54]: View transitions were messed up. when you brought up 3dot menu the screen time vc disappeared.
This has been discussed with @kylehickinson

File: Sources/Brave/Frontend/Browser/BrowserViewController.swift (hunk 3 above)
Comment (iccub) [reply Sources/Brave/Frontend/Browser/BrowserViewController.swift]: let me try

Comment (iccub) [reply Sources/Brave/Frontend/Browser/BrowserViewController.swift]: Looks like i do not even need it

**Review by soner-yuksel — APPROVED** [review]

**Review by Brandon-T — APPROVED** [review]

File: Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ScreenTime.swift [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import Preferences
+
+extension BrowserViewController {
+  /// Updates the url of the screen time controller. Keep in mind a single screen time controller is used for every window.
+  /// Multiple instances of it seem to crash the app.
+  /// There is also one hack required: STWebpageController breaks if you pass scheme other than http or https,
+  /// it will not block anything for the rest of its lifecycle. Our internal urls have to be bridged to an empty https url.
+  func updateScreenTimeUrl(_ url: URL?) {
+    guard let url = url, (url.scheme == "http" || url.scheme == "https") else {
+      screenTimeViewController?.url = URL(string: "https://about:blank")
```
Comment (stoletheminerals) [inline Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ScreenTime.swift:16]: Do we, for the sake of the screen time control, change all internal urls to `https://about:blank` here? In the case if a user would enable time limit for `https://about:blank`, will it break all our interstitials?

Comment (iccub) [reply Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ScreenTime.swift:16]: It does not break but the screen limit blinks for a split second.
We will look into trying to fix this

Comment (iccub) [reply Sources/Brave/Frontend/Browser/BrowserViewController/BVC+ScreenTime.swift:16]: But it seems a minor issue, people will not put this website as their screen limit obviously

**Pull request merged**